keeper/accounting.go

```diff
@@ -0,0 +1,395 @@
+func (k *Keeper) updateVaultsV2AccountingWithCursor(ctx context.Context, maxPositions uint32) (*AccountingResult, error) {
+	navInfo, err := k.GetVaultsV2NAVInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if navInfo.CurrentNav.IsNil() {
+		return nil, fmt.Errorf("current NAV is not set")
+	}
+
+	state, err := k.GetVaultsV2VaultState(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	totalShares, err := k.GetVaultsV2TotalShares(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cursor, err := k.GetVaultsV2AccountingCursor(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Determine if we need to start a new accounting session
+	needsInit := !cursor.InProgress ||
+		cursor.AccountingNav.IsNil() ||
+		!cursor.AccountingNav.Equal(navInfo.CurrentNav) ||
+		!cursor.AccountingNavTimestamp.Equal(navInfo.LastUpdate)
+
+	if needsInit {
+		// If accounting is in progress with a different NAV, we have a problem
+		// We cannot start new accounting while old accounting is incomplete
+		// as this would leave some users with old NAV and some with new NAV
+		if cursor.InProgress {
+			return nil, fmt.Errorf(
+				"accounting already in progress for NAV %s (started at %s, %d/%d positions processed). "+
+					"Cannot start new accounting for NAV %s until current session completes. "+
+					"Continue calling this message to complete the current session",
+				cursor.AccountingNav.String(),
+				cursor.StartedAt.String(),
+				cursor.PositionsProcessed,
+				cursor.TotalPositions,
+				navInfo.CurrentNav.String(),
+			)
+		}
+
+		// Count total positions
+		totalPositions := uint64(0)
+		if err := k.IterateVaultsV2UserPositions(ctx, func(address types.AccAddress, position vaultsv2.UserPosition) (bool, error) {
+			totalPositions++
+			return false, nil
+		}); err != nil {
+			return nil, err
+		}
+
+		// Initialize new accounting session
+		cursor = vaultsv2.AccountingCursor{
+			LastProcessedUser:        "",
+			AccountingNav:            navInfo.CurrentNav,
+			AccountingNavTimestamp:   navInfo.LastUpdate,
+			PositionsProcessed:       0,
+			TotalPositions:           totalPositions,
+			InProgress:               true,
+			StartedAt:                k.header.GetHeaderInfo(ctx).Time,
+			AccumulatedResidual:      sdkmath.ZeroInt(),
+		}
+	}
+
+	// When there are no active shares, reset accrued yield and mirror NAV
+	if totalShares.IsZero() {
+		return k.accountingWithZeroShares(ctx, navInfo, state, cursor)
+	}
+
+	// Perform cursor-based accounting
+	return k.accountingWithCursor(ctx, navInfo, state, totalShares, cursor, maxPositions)
+}
+
+// accountingWithZeroShares handles the case where there are no active shares.
+func (k *Keeper) accountingWithZeroShares(
+	ctx context.Context,
+	navInfo vaultsv2.NAVInfo,
+	state vaultsv2.VaultState,
+	cursor vaultsv2.AccountingCursor,
+) (*AccountingResult, error) {
+	positionsProcessed := uint32(0)
+	maxBatch := uint32(100) // Process in batches when resetting yield
```

Comment:
Is there a reason it's hardcoded instead of passing in the `maxPositions` parameter from the caller?